### PR TITLE
tests/lib: support for sequence forming assertions in fakestore code

### PR DESCRIPTION
**Pull request by Meulengracht**
This adds support for fetching sequence forming assertions, and teaches fakestore how to return these.

### CODE REVIEW

File: tests/lib/fakestore/store/store.go [hunk 1]
```diff
@@ -739,49 +740,109 @@ func (s *Store) retrieveAssertion(bs asserts.Backstore, assertType *asserts.Asse
...
+func (s *Store) retrieveAssertionWrapper(bs asserts.Backstore, assertType *asserts.AssertionType, keyParts []string, values url.Values) (asserts.Assertion, error) {
+
+	pk := keyParts
+	if assertType.SequenceForming() {
+		seq, err := s.sequenceFromQueryValues(values)
+		if err != nil {
+			return nil, err
+		}
+
+		// If no sequence key is present, then we use a special lookup to
```
Comment (ernestl) [inline tests/lib/fakestore/store/store.go:801]: Looks like it should perhaps be 'No sequence key or sequence key "latest"' will result in the -1 special lookup?

File: tests/lib/fakestore/store/store.go [hunk 2]
```diff
@@ -739,49 +740,109 @@ func (s *Store) retrieveAssertion(bs asserts.Backstore, assertType *asserts.Asse
 	return a, err
 }
 
-func (s *Store) assertionsEndpoint(w http.ResponseWriter, req *http.Request) {
-	assertPath := strings.TrimPrefix(req.URL.Path, "/v2/assertions/")
+func (s *Store) retrieveSequenceFormingAssertion(bs asserts.Backstore, assertType *asserts.AssertionType, sequenceKey []string, sequence int) (asserts.Assertion, error) {
+	if sequence > 0 {
```
Comment (ernestl) [inline tests/lib/fakestore/store/store.go:744]: With the current use case for this function, this if block is dead code which made me question the intent.

From documentation `SeqFormingAssertion` "sequence <= 0 we query for the latest sequence",
which seems to suggest the sequence (-1 or 0) does not play a functional role and this function always retrieves the latest sequence. 

Perhaps this function can be simplified and called something like `retrieveLatestSequenceFormingAssertion` without passing `sequence` and using hardcoded -1 for 'fallback.SeqFormingAssertion'? This then relies on the caller to only call when applicable (sequence > 0) which is already the case.

File: tests/lib/fakestore/store/store_test.go [hunk 3]
```diff
@@ -478,6 +495,45 @@ func (s *storeTestSuite) TestAssertionsEndpointFromAssertsDir(c *C) {
...
+	body, err := ioutil.ReadAll(resp.Body)
+	c.Assert(err, IsNil)
+	c.Check(string(body), Equals, exampleValidationSet)
+}
+
+func (s *storeTestSuite) TestAssertionsEndpointSequenceAssertionLatest(c *C) {
+	err := ioutil.WriteFile(filepath.Join(s.store.assertDir, "base-set.validation-set"), []byte(exampleValidationSet), 0655)
+	c.Assert(err, IsNil)
+
+	resp, err := s.StoreGet(`/v2/assertions/validation-set/16/canonical/base-set?sequence=latest`)
```
Comment (ernestl) [inline tests/lib/fakestore/store/store_test.go:516]: Suggest adding additional case where 'sequenceFromQueryValues' returns 0 which should also result in latest.

File: tests/lib/fakestore/store/store.go (hunk 1 above)
Comment (Meulengracht) [reply tests/lib/fakestore/store/store.go]: This makes a lot of sense, thank you!

File: tests/lib/fakestore/store/store.go (hunk 2 above)
Comment (Meulengracht) [reply tests/lib/fakestore/store/store.go]: Great point, there was no point in adding that case currently as we never hit it. I have simplified the function as suggested.

File: tests/lib/fakestore/store/store_test.go (hunk 3 above)
Comment (Meulengracht) [reply tests/lib/fakestore/store/store_test.go:516]: Done, nice catch!

File: tests/lib/fakestore/store/store.go [hunk 4]
```diff
@@ -739,49 +740,101 @@ func (s *Store) retrieveAssertion(bs asserts.Backstore, assertType *asserts.Asse
...
+	if val, ok := values["sequence"]; ok {
+		// special case value of 'latest', in that case
+		// we return -1 to indicate we want the newest
+		if val[0] != "latest" {
+			seq, err := strconv.Atoi(val[0])
+			if err != nil {
+				return -1, fmt.Errorf("cannot parse sequence %s: %v", val[0], err)
+			}
+			// XXX: Treat values 0 and below as error? Anyone know the official store
+			// behavior on this?
```
Comment (pedronis) [inline tests/lib/fakestore/store/store.go:761]: I think it errors:

sequence param must be a positive integer or "latest"

**Review by pedronis — COMMENTED** [review]
thanks

**Review by pedronis — APPROVED** [review]

File: tests/lib/fakestore/store/store.go (hunk 4 above)
Comment (Meulengracht) [reply tests/lib/fakestore/store/store.go]: Thank you for the insight, changed this and the unit test to error on 0 and negative values

**Pull request merged**